File: java/src/main/java/PvaultGettingStarted.java
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import com.piiano.vault.client.openapi.*;
import com.piiano.vault.client.openapi.model.Collection;
import com.piiano.vault.client.openapi.model.*;

import java.util.*;

import static java.util.Collections.emptyList;
import static java.util.Collections.emptySet;

public class PvaultGettingStarted {

    public static final String COLLECTION_NAME = "customers";
    public static final String HEALTH_PASS = "pass";
    public static final String JSON = "json";
    public static final String APP_FUNCTIONALITY_REASON = "AppFunctionality";
    public static final String NO_ADHOC_REASON = null;
    public static final Set<String> NO_OPTIONS = emptySet();
    public static final String USE_DEFAULT_TTL = "";
    public static final String UNSAFE_OPTION = "unsafe"; // fetch all the properties
    public static final String NO_TRANSACTION_ID = null; // Transaction ID is only relevant for advanced usage
    public static final int PVAULT_ADDRESS = 8123;

    public void run() throws Exception {

        print("\n\n== Steps 1 + 2: Connect to Piiano vault and check status ==\n\n");
        ApiClient pvaultClient = getApiClient();
        checkPvaultStatus(pvaultClient);

        print("\n\n== Step 3: Create a collection ==\n\n");
        CollectionsApi collectionsApi = new CollectionsApi(pvaultClient);
        CollectionPropertiesApi collectionpropapi = new CollectionPropertiesApi(pvaultClient);
        // deleteCollection(collectionsApi);
        verifyNoCollections(collectionpropapi);
        createCollection(collectionsApi);

        print("\n\n== Step 4: Add data ==\n\n");
        ObjectsApi objectsApi = new ObjectsApi(pvaultClient);
        List<UUID> customerIds = addData(objectsApi);

        print("\n\n== Step 5: Tokenize data ==\n\n");
        TokensApi tokensApi = new TokensApi(pvaultClient);
        String token = tokenizeData(customerIds.get(0), tokensApi);

        detokenizeToken(tokensApi, token);

        print("\n\n== Step 6: Query your data ==\n\n");
        queryAllObjectsWithPageSize(objectsApi);
        queryPropertiesOfObjectsById(objectsApi, customerIds.get(0));
        getTransformedPropertiesOfObjects(objectsApi, customerIds.get(0));

        print("\n\n== Step 7: Delete data ==\n\n");
        deleteToken(tokensApi, token);
        deleteObject(objectsApi, customerIds.get(0));

        print("Done!\n");
    }

    private static void checkPvaultStatus(ApiClient pvaultClient) throws ApiException {

        SystemApi systemApi = new SystemApi(pvaultClient);
        String controlStatus = systemApi.controlHealth().getStatus();
        String dataStatus = systemApi.dataHealth().getStatus();

        print("control status: ", controlStatus);
        print("data status: ", dataStatus);
        assert HEALTH_PASS.equals(controlStatus) && HEALTH_PASS.equals(dataStatus);
    }

    private static void deleteCollection(CollectionsApi collectionApi) {
        print("Clearing the collection from previous runs");
        try {
            collectionApi.deleteCollection(COLLECTION_NAME);
        } catch (ApiException e) {
            // test collection wasn't exist, continue test..
        }
    }

    // for safety reasons refuse to run if the collection already exists
    private static void verifyNoCollections(CollectionPropertiesApi collectionpropApi) throws ApiException {
        print("Verifying the test collection is not present");
        try {
            List<Property> mp =
                    collectionpropApi.listCollectionProperties(COLLECTION_NAME, NO_OPTIONS);
            throw new RuntimeException("Collection " + COLLECTION_NAME + " already exists.\n" +
                    "Recreate the Vault from scratch or uncomment deleteCollection()" +
                    " in this code. Bailing out.\n");
        } catch (ApiException e) {
            if (e.getCode() == 404) {
                print("Collection " + COLLECTION_NAME + " not found. Will create it");
            } else {
                throw new ApiException(e);
            }
        }
    }

    private static void createCollection(CollectionsApi collectionApi) throws ApiException {
        // Note: Adding a collection with pvschema is not supported in the SDK
        // Throughout this code we will use JSON exclusively.

        Collection collection = new Collection();
        collection.setName(COLLECTION_NAME);
        collection.setType(Collection.TypeEnum.PERSONS);

        collection.addPropertiesItem(
                buildProperty("ssn", "SSN", "Social security number",
                        true, false, true, false));

        collection.addPropertiesItem(
                buildProperty("email", "EMAIL", "EMAIL",
                        false, false, true, false));

        collection.addPropertiesItem(
                buildProperty("phone_number", "PHONE_NUMBER", "PHONE_NUMBER",
                        false, true, true, false));

        collection.addPropertiesItem(
                buildProperty("zip_code_us", "ZIP_CODE_US", "ZIP_CODE_US",
                        false, true, true, false));

        collectionApi.addCollection(collection, JSON, NO_OPTIONS);

        // Check the collections has been added.
        collection = collectionApi.getCollection(COLLECTION_NAME, JSON, NO_OPTIONS);
        assert collection != null;
        print("collection: ", collection.toString());
    }

    private static List<UUID> addData(ObjectsApi objectsApi) throws ApiException {

        List<UUID> customerIds = new ArrayList<>();

        Map<String, Object> objectDetails = buildObjectDetails(
                "123-12-1234", "john@somemail.com",
                "+1-121212123", "12345");
        UUID customer1ID = objectsApi.addObject(COLLECTION_NAME, APP_FUNCTIONALITY_REASON, objectDetails,
                NO_ADHOC_REASON, false, USE_DEFAULT_TTL).getId();
        customerIds.add(customer1ID);
        print("customer1 ID: ", customer1ID.toString());

        objectDetails = buildObjectDetails(
                "123-12-1235", "mary@somemail.com",
                "+1-121212124", "12345");
        UUID customer2ID = objectsApi.addObject(COLLECTION_NAME, APP_FUNCTIONALITY_REASON, objectDetails,
                NO_ADHOC_REASON, false, USE_DEFAULT_TTL).getId();
        customerIds.add(customer2ID);
        print("customer2 ID: ", customer2ID.toString());

        objectDetails = buildObjectDetails(
                "123-12-1236", "eric@somemail.com",
                "+1-121212125", "12345");

        UUID customer3ID = objectsApi.addObject(COLLECTION_NAME, APP_FUNCTIONALITY_REASON, objectDetails,
                NO_ADHOC_REASON, false, USE_DEFAULT_TTL).getId();
        customerIds.add(customer3ID);
        print("customer3 ID: ", customer3ID.toString());

        return customerIds;
    }

    private static String tokenizeData(UUID id, TokensApi tokensApi) throws ApiException {

        TokenizeRequest tokenizeRequest = new TokenizeRequest();
        InputObject object = new InputObject();
        object.setId(id);
        object.setFields(null);
        tokenizeRequest.setObject(object);
        tokenizeRequest.addPropsItem("email");
        tokenizeRequest.setType(TokenType.POINTER);
        tokenizeRequest.setTags(ImmutableList.of("token_tag"));

        String token = tokensApi.tokenize(COLLECTION_NAME, APP_FUNCTIONALITY_REASON, List.of(tokenizeRequest),
                USE_DEFAULT_TTL, NO_TRANSACTION_ID, NO_ADHOC_REASON, false).get(0).getTokenId();
        print("Token: ", token);
        return token;
    }

    private static void detokenizeToken(TokensApi tokensApi, String token) throws ApiException {

        String returnedEmail = tokensApi.detokenize(COLLECTION_NAME, APP_FUNCTIONALITY_REASON, emptyList(),
                        emptyList(), ImmutableList.of(token), emptyList(), NO_OPTIONS, NO_ADHOC_REASON, false)
                .get(0).getFields().get("email").toString();
        assert "john@somemail.com".equals(returnedEmail);
    }

    private static void queryAllObjectsWithPageSize(ObjectsApi objectsApi) throws ApiException {

        ObjectFieldsPage objectIdsPage =
                objectsApi.listObjects(COLLECTION_NAME, APP_FUNCTIONALITY_REASON, NO_ADHOC_REASON,
                        false, 1, "", "", emptyList(), ImmutableSet.of(UNSAFE_OPTION), null);

        assert objectIdsPage.getResults().size() == 1;
        Map<String, Object> searchResult = objectIdsPage.getResults().get(0);

        print("object retrieved by search: ", searchResult.toString());
        assert "john@somemail.com".equals(searchResult.get("email"));
        assert "123-12-1234".equals(searchResult.get("ssn"));
        assert "+1121212123".equals(searchResult.get("phone_number"));
        assert "12345".equals(searchResult.get("zip_code_us"));
    }

    private static void queryPropertiesOfObjectsById(ObjectsApi objectsApi, UUID id) throws ApiException {

        ObjectFieldsPage objectIdsPage = objectsApi.listObjects(COLLECTION_NAME, APP_FUNCTIONALITY_REASON,
                NO_ADHOC_REASON, false, null, "", "", ImmutableList.of(id),
                NO_OPTIONS, ImmutableList.of("ssn"));

        assert objectIdsPage.getResults().size() == 1;
        Map<String, Object> searchResult = objectIdsPage.getResults().get(0);
        assert searchResult.size() == 1;

        print("ssn property retrieved by id: ", searchResult.toString());
        assert "123-12-1234".equals(searchResult.get("ssn"));
    }

    private static void getTransformedPropertiesOfObjects(ObjectsApi objectsApi, UUID id) throws ApiException {

        ObjectFieldsPage objectIdsPage = objectsApi.listObjects(COLLECTION_NAME, APP_FUNCTIONALITY_REASON,
                NO_ADHOC_REASON, false, null, "", "", ImmutableList.of(id),
                NO_OPTIONS, ImmutableList.of("ssn.mask", "email.mask", "phone_number.mask"));

        assert objectIdsPage.getResults().size() == 1;
        Map<String, Object> searchResult = objectIdsPage.getResults().get(0);

        print("transformed propertied retrieved: ", searchResult.toString());
        assert "j***@somemail.com".equals(searchResult.get("email.mask"));
        assert "***-**-1234".equals(searchResult.get("ssn.mask"));
        assert "*******2123".equals(searchResult.get("phone_number.mask"));
    }

    private static void deleteToken(TokensApi tokensApi, String token) throws ApiException {

        tokensApi.deleteTokens(COLLECTION_NAME, APP_FUNCTIONALITY_REASON, emptyList(), emptyList(),
                ImmutableList.of(token), null, NO_OPTIONS, NO_ADHOC_REASON, false);
    }

    private static void deleteObject(ObjectsApi objectsApi, UUID id) throws ApiException {

        objectsApi.deleteObjectById(COLLECTION_NAME, id, APP_FUNCTIONALITY_REASON,
                NO_OPTIONS, NO_ADHOC_REASON, false);
    }

    private static ApiClient getApiClient() {

        // Create configuration, bearer auth and client API
        ApiClient pvaultClient = Configuration.getDefaultApiClient();
        pvaultClient.setBasePath("http://localhost:" + PVAULT_ADDRESS);
        pvaultClient.setBearerToken("pvaultauth");
        pvaultClient.addDefaultHeader("Content-Type", "application/json");
        return pvaultClient;
    }

    private static Property buildProperty(
            String name, String dataTypeName, String description,
            boolean isUnique, boolean isNullable, boolean isEncrypted, boolean isIndex) {

        Property property = new Property();
        property.setName(name);
        property.setDataTypeName(dataTypeName);
        property.setDescription(description);
        property.setIsUnique(isUnique);
        property.isNullable(isNullable);
        property.setIsEncrypted(isEncrypted);
        property.isIndex(isIndex);
        return property;
    }

    private static Map<String, Object> buildObjectDetails(
            String ssn, String email, String phoneNumber, String zipCodeUS) {

        Map<String, Object> objectDetails = new HashMap<>();
        objectDetails.put("ssn", ssn);
        objectDetails.put("email", email);
        objectDetails.put("phone_number", phoneNumber);
        objectDetails.put("zip_code_us", zipCodeUS);
        return objectDetails;
    }

    private static void print(String... args) {
        for (String arg : args) {
            System.out.print(arg);
        }
        System.out.println();
    }

    public static void main(String[] args) {
        PvaultGettingStarted pvaultGettingStarted = new PvaultGettingStarted();
        try {
            pvaultGettingStarted.run();
        } catch (ApiException apiException) {
            if (apiException.getMessage().contains("java.net.ConnectException")) {
                print(apiException.getMessage() + "\n\nIs the Vault running?\n");
            } else {
                throw new RuntimeException(apiException);
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: typescript/src/PvaultGettingStarted.ts
import {
  Collection,
  CollectionsApi,
  ModelObject,
  ObjectsApi,
  QueryToken,
  SystemApi,
  TokenizeRequest,
  TokensApi,
  TokenType,
  CollectionTypeEnum,
  Configuration,
} from "../vault_typescript_sdk";

interface CustomersCollectionObject extends ModelObject {
  ssn: string;
  email: string;
  phone_number?: string;
  zip_code_us?: string;
}

interface RemoteCustomersCollectionObject extends Partial<CustomersCollectionObject> {
  id?: string;
}

class PvaultGettingStarted {
  constructor(private host: string, private port: string, private apiToken: string, private conf: any) {}
  public async run() {
    console.log("\n\n== Steps 1 + 2: Connect to Piiano vault and check status ==\n\n");
    this.setupApiClient();

    await this.prepareVault();

    console.log("\n\n== Step 3: Create a collection ==\n\n");
    const customersCollectionSchema: Collection = {
      name: "customers",
      type: CollectionTypeEnum.Persons,
      properties: [
        {
          name: "ssn",
          data_type_name: "SSN",
          is_nullable: true,
          description: "Social Security Number",
        },
        {
          name: "email",
          data_type_name: "EMAIL",
        },
        {
          name: "phone_number",
          data_type_name: "PHONE_NUMBER",
          is_nullable: true,
        },
        {
          name: "zip_code_us",
          data_type_name: "ZIP_CODE_US",
          is_nullable: true,
        },
      ],
    };
    const customersCollection = await this.createCollection(customersCollectionSchema, "json");

    console.log("\n\n== Step 4: Add data ==\n\n");
    const objects: CustomersCollectionObject[] = [
      {
        ssn: "123-12-1234",
        email: "john@somemail.com",
        phone_number: "+1-121212123",
        zip_code_us: "12345",
      },
      {
        ssn: "123-12-1235",
        email: "mary@somemail.com",
        phone_number: "+1-121212123",
        zip_code_us: "12345",
      },
      {
        ssn: "123-12-1236",
        email: "eric@somemail.com",
        phone_number: "+1-121212123",
        zip_code_us: "12345",
      },
    ];

    const customers = await this.addObjectsToCollection(customersCollection.data, objects);

    console.log("\n\n== Step 5: Tokenize data ==\n\n");
    const tokenRequest: TokenizeRequest = {
      type: TokenType.Pointer,
      props: ["email"],
      object: {
        id: customers[0].id,
      },
    };

    const searchTokenRequest: QueryToken = {
      object_ids: [customers[0].id!],
    };

    const tokenId = await this.tokenizeObject(
      customersCollection.data,
      customers[0],
      tokenRequest,
      searchTokenRequest
    );

    console.log("\n\n== Step 6: Query your data ==\n\n");
    await this.queryObjects(customersCollection.data, customers, customers[0], "ssn");

    console.log("\n\n== Step 7: Delete data ==\n\n");
    await this.deleteObject(customersCollection.data, tokenId, customers[0].id!, searchTokenRequest);

    console.log("\n\nDone!\n");
  }

  public setupApiClient() {
    console.log("Setting up API client connection...");

    // Set up API client
    this.conf = new Configuration({
      basePath: `${this.host}:${this.port}`,
      accessToken: this.apiToken
    })
  }

  public async prepareVault() {
    console.log("Preparing vault...");
    console.log("\tChecking vault is empty...");

    // Fetch all collections
    const collection_client = new CollectionsApi(this.conf)
    const system_client = new SystemApi(this.conf)

    const collections = await collection_client.listCollections();
    if (collections.data.length > 0) throw new Error('Vault is not empty. Please run this script on a clean instance.');

    console.log("\tChecking health of vault...");

    // Check vault health
    const health = await system_client.controlHealth();

    if (health.data.status !== "pass") throw new Error("Health check failed.");
  }

  public async createCollection(schema: Collection, format: "json" | "pvschema" = "json") {
    console.log(`Adding collection ${schema.name}...`);
    const collection_client = new CollectionsApi(this.conf)
    // Create collection
    const collection = await collection_client.addCollection(schema, format);

    // Fetch collection by name
    const newCollection = await collection_client.getCollection(collection.data.name);

    if (!newCollection) throw new Error("Collection not found.");
    console.log(
      `Collection details:\n\ttype: ${collection.data.type}\n\tname: ${collection.data.name}\n\tcreation time: ${collection.data.creation_time}`
    );

    return collection;
  }

  public async addObjectsToCollection(
    collection: Collection,
    objects: CustomersCollectionObject[]
  ) {
    const _ = undefined;
    console.log(`Adding objects to collection ${collection.name}...`);
    const object_client = new ObjectsApi(this.conf)
    // Add objects to collection
    const resultObjects = await object_client.addObjects(
          collection.name,
          "AppFunctionality",
          objects,
    );

    if (!resultObjects.data.ok) throw new Error("Failed to add objects to the collection.");
    console.log(
      `Added objects with the following ID's to collection ${
        collection.name
      }:${resultObjects.data.results.map((item) => `\n\t${item.id}`)}`
    );

    console.log(
      `Searching for object by email (john@somemail.com) in collection ${collection.name}...`
    );

    // Search for objects in collection by email
    //
    // Here you need to specify the last parameter of the searchObjects function,
    // however, you might want to skip the ones in between, hence the usage of _ = undefined
    // It would be a good idea to wrap these generated API functions in your own functions
    // based on your desired usage
    const searchResult = await object_client.searchObjects(
      collection.name,
      "AppFunctionality",
      {
        match: { email: "john@somemail.com" },
      },
      _,
      _,
      _,
      _,
      _,
      _,
      ["id", "email"]
    );

    if (searchResult.data.results.length === 0)
      throw new Error("Failed to search objects in the collection.");
    console.log(
      `Found the following objects in collection ${collection.name}:${searchResult.data.results.map(
        (item: RemoteCustomersCollectionObject) => `\n\t${JSON.stringify(item)}`
      )}`
    );

    return resultObjects.data.results;
  }

  public async tokenizeObject(
    collection: Collection,
    object: RemoteCustomersCollectionObject,
    tokenRequest: TokenizeRequest,
    searchTokenRequest: QueryToken
  ) {
    console.log(`Tokenizing object with ID ${object.id} in collection ${collection.name}...`);
    const token_client = new TokensApi(this.conf)
    // Tokenize object
    const tokens = await token_client.tokenize(
      collection.name,
      "AppFunctionality",
      [
        tokenRequest
      ]
    );

    if (!tokens.data[0]) throw new Error("Failed to tokenize object.");
    console.log(
      `Tokenized object with ID ${object.id} in collection ${collection.name} with token ID ${tokens.data[0].token_id}.`
    );

    // Search for token
    const searchTokens = await token_client.searchTokens(
      collection.name,
      "AppFunctionality",
      searchTokenRequest
    );

    if (!searchTokens.data[0])
      throw new Error(`Failed to find a token for the object with ID ${object.id}.`);
    console.log(
      `Found token with ID ${searchTokens.data[0].token_id} in collection ${collection.name}.`
    );

    return tokens.data[0].token_id;
  }

  public async queryObjects(
    collection: Collection,
    objects: RemoteCustomersCollectionObject[],
    object: RemoteCustomersCollectionObject,
    prop: string
  ) {
    const _ = undefined;

    console.log(`Querying paginated objects in collection ${collection.name}...`);
    const object_client = new ObjectsApi(this.conf)
    // Query all objects with pagination enabled
    const allObjects = await object_client.listObjects(
      collection.name,
      "AppFunctionality",
      _,
      _,
      _,
      _,
      _,
      _,
      ["unsafe"]
    );

    if (allObjects.data.results.length === 0)
      throw new Error("Failed to query objects in the collection.");
    console.log(
      `Found the following objects in collection ${collection.name}:${allObjects.data.results.map(
        (item: RemoteCustomersCollectionObject) => `\n\t${JSON.stringify(item)}`
      )}`
    );
    console.log(`Pagination:\n\t${JSON.stringify(allObjects.data.paging)}`);

    console.log(`\nQuerying for a specific object by ID in collection ${collection.name}...`);

    // Query only one object with a specific prop
    const objectWithProp = await object_client.listObjects(
      collection.name,
      "AppFunctionality",
        _,
        _,
        _,
        _,
        _,
        [object.id!],
        _,
        [prop]
    );

    if (objectWithProp.data.results.length === 0)
      throw new Error(`Failed to query object-${object.id} in the collection with ${prop} prop.`);
    console.log(
      `Found the following objects in collection ${collection.name}:${objectWithProp.data.results.map(
        (item: RemoteCustomersCollectionObject) => `\n\t${JSON.stringify(item)}`
      )}`
    );

    console.log(
      `\nQuerying all the details for a specific object by ID in collection ${collection.name}...`
    );

    // Query only one object with all the props
    const objectsWithAllProps = await object_client.listObjects(
      collection.name,
      "AppFunctionality",
        _,
        _,
        _,
        _,
        _,
        [object.id!],
        _,
        collection.properties.map((item) => item.name)
    );
    if (objectsWithAllProps.data.results.length === 0)
      throw new Error(`Failed to query object-${object.id} in the collection.`);

    console.log(
      `Found the following objects in collection ${
        collection.name
      }:${objectsWithAllProps.data.results.map(
        (item: RemoteCustomersCollectionObject) => `\n\t${JSON.stringify(item)}`
      )}`
    );

    // Query only one object with all the props masked
    const objectsWithMasks = await object_client.listObjects(
        collection.name,
      "AppFunctionality",
        _,
        _,
        _,
        _,
        _,
        [object.id!],
        _,
        collection.properties
          .map((item) => `${item.name}.mask`)
          .filter((item) => item !== "zip_code_us.mask")
    );
    if (objectsWithMasks.data.results.length === 0)
      throw new Error(`Failed to query object-${object.id} in the collection.`);

    console.log(
      `Found the following objects in collection ${collection.name}:${objectsWithMasks.data.results.map(
        (item: RemoteCustomersCollectionObject) => `\n\t${JSON.stringify(item)}`
      )}`
    );
  }

  public async deleteObject(
    collection: Collection,
    tokenId: string,
    objectId: string,
    searchTokenRequest: QueryToken
  ) {
    const _ = undefined;
    console.log(`Deleting token with ID ${tokenId} in collection ${collection.name}...`);

    const token_client = new TokensApi(this.conf)
    const object_client = new ObjectsApi(this.conf)

    // Delete token
    await token_client.deleteTokens(
      collection.name,
      "AppFunctionality",
      _,
      _,
      [tokenId],
    );
    const tokens = await token_client.searchTokens(
          collection.name,
          "AppFunctionality",
          searchTokenRequest,
    );
    if (tokens.data.length !== 0)
      throw new Error(
        `Failed to delete token with ID ${tokenId} in collection ${collection.name}.`
      );

    console.log(`Deleting object with ID ${objectId} in collection ${collection.name}...`);

    // Delete object
    const deleteResult = await object_client.deleteObjectById(
      collection.name,
      objectId,
    "AppFunctionality",
    );
    try {
      await object_client.getObjectById(
        collection.name,
        objectId,
          "AppFunctionality",
        _,
        _,
        _,
        ["unsafe"]
      );
    } catch (error: any) {
      if (error.response.status !== 404)
        throw new Error(
          `Failed to delete object with ID ${objectId} in collection ${collection.name}.`
        );
    }
  }
}

// Initialize the Pvault instance
const pvaultGettingStarted = new PvaultGettingStarted(
  process.env.PVAULT_HOST || "http://localhost",
  process.env.PVAULT_PORT || "8123",
  process.env.PVAULT_TOKEN || "pvaultauth",
    null
);

// Run the demo
pvaultGettingStarted.run().catch((error) => {
  console.error(error);
});
